Feat: roman numerals instead of arabic ones for companies and refactor of code

felipeStouvenel opened the pull request:
# Feat/fixes for scr_management and scr_ui_manage

## List of changes

- Changed company name to display roman numerals (I ; II , III, IV, V, VI, VII, VIII, IX, X)
- Refactored all repeat functions in scr_management and scr_ui_manage
- Re did logic to assign hp values to non marine units
- Added a key list with values for each non marine unit to have an hp already setup
- Improved some if loops
- Added scr_roman_numerals which returns the roman numerals
- Implemented the new script into existing code
- Removed all repeat functions from company_view
- Refactored ui_advisors
- Added arrays for recruitment_rates and recruitement_pace inside ui_advisors
- Refactored ui_settings
- Refactored obj_massequip_step script
- Changed all company numeric values to use Roman values
- Refactored obj_shop draw_0 script
- Refactored obj_popup draw_0 script
- Refactored scr_convert_to_string script
- Fixed some bugs from GM2 (red ones) mainly extra parameters on some draw functions and other minor details

GalacticChimp commented:
converted to draft, as this is a WIP. Once you are satisfied with your changes, convert to normal PR

felipeStouvenel commented:
Finished the clean up code and did some logic improvements where possible

GalacticChimp commented:
please add a summary of changes in the description

GalacticChimp commented on scripts/scr_ui_manage/scr_ui_manage.gml:
```diff
@@ -1,46 +1,69 @@
+	var normal_hp=true;
+	
+	// Declare non marine roles here
+	var non_marine_roles=[
```
we should really switch to string ids (e.g. sister_of_battle). But that's for the future.

GalacticChimp commented on scripts/scr_ui_manage/scr_ui_manage.gml:
```diff
@@ -1,46 +1,69 @@
 function scr_ui_manage() {
+	
+	var romanNumerals=[
```
I'd extract this definition to a separate script, so we won't have to repeat it in multiple scripts

GalacticChimp commented:
missing spot:

![image](https://github.com/ZAB909/ChapterMaster/assets/63882831/1434b843-7d77-4a18-9f97-59d978c8b416)

GalacticChimp commented:
also missing in Chapter Settings -> Astartes Role Settings -> 

![image](https://github.com/ZAB909/ChapterMaster/assets/63882831/df85eaca-6495-49a0-a472-6f982fd26e1e)

GalacticChimp reviewed (changes requested):
overall fine, just please extract the roman numerals and add numerals to the remaining views

felipeStouvenel commented:
Fixed remaining issues and refactored some scripts
![image](https://github.com/ZAB909/ChapterMaster/assets/142210479/be64fd0c-b861-4764-8fd0-806b03c06994)
![image](https://github.com/ZAB909/ChapterMaster/assets/142210479/dd12986e-f1c9-48b6-bd97-dc460b5d1d01)

GalacticChimp commented on scripts/scr_company_view/scr_company_view.gml:
```diff
@@ -46,30 +73,31 @@ function scr_company_view(argument0) {
-	                        if (sel_uni[i]="") and (op=0) then op=i;
-	                        if (sel_uni[i]=ma_role[v]) then go=1;
-	                    }
+	                    for (var j=-1; j<=20;j++) {
```
why start from -1?

GalacticChimp commented on scripts/scr_ui_settings/scr_ui_settings.gml:
```diff
@@ -1,30 +1,42 @@
 function scr_ui_settings() {
 
+	var romanNumerals;
+	romanNumerals= scr_roman_numerals();
```
next time please remember to declare and init on the same line, makes no sense to split it

GalacticChimp commented on objects/obj_mass_equip/Step_0.gml:
```diff
@@ -1,15 +1,17 @@
-    var co,i;co=-1;i=0;
-    
-    repeat(11){co+=1;i=0;
+    for(var co=0, i=0; co<11; co++){
```
are you sure this will reset `i` to 0 after each loop pass?

felipeStouvenel replied:
it does on line 9, deleted i=0 as its unnecessary now

GalacticChimp replied:
line 9 is just an if. you need to remove the i=0 from line 8

GalacticChimp commented:
clicking a Company button crashes the game:

![image](https://github.com/ZAB909/ChapterMaster/assets/63882831/fb830553-00c1-4770-9c1d-d72a9310fc45)

GalacticChimp commented on objects/obj_mass_equip/Step_0.gml:
```diff
@@ -1,15 +1,17 @@
+    for(var co=0, i=0; co<11; co++){
         if (role_number[co]>0){
-            repeat(300){i+=1;
+			for(i=0; i<300; i++){
```
also, this should start from 1. Previously, the first pass through the `repeat(300){...` incremented `i` at the start, so it was always at least 1 in the loop

GalacticChimp commented on objects/obj_mass_equip/Step_0.gml:
```diff
@@ -72,81 +86,159 @@ if (engage=true){
-        repeat(300){i+=1;
+	
+	for(var co=0; co<11; co++){
+		for(var i=0; i<300; i++){
```
should also start with `i = 0` and `i <= 300`

GalacticChimp commented on scripts/scr_company_view/scr_company_view.gml:
```diff
@@ -224,12 +255,12 @@ function scr_company_view(argument0) {
+	last_vehicle=0;
 
-	repeat(100){
+	for (var i = 0; i < 100; i++){
```
same, `var i = 1; i <= 100`

GalacticChimp commented on scripts/scr_company_view/scr_company_view.gml:
```diff
@@ -1,33 +1,60 @@
-	    ma_race[i]=0;ma_loc[i]="";ma_name[i]="";ma_role[i]="";ma_gear[i]="";ma_mobi[i]="";ma_wep1[i]="";
-	    ma_wep2[i]="";ma_armour[i]="";ma_health[i]=100;ma_chaos[i]=0;ma_exp[i]=0;ma_promote[i]=0;ma_god[i]=0;
-	}i=0;
+	for (var i = 0; i < 499; i++){
```
same, `var i = 1; i <= 499`

GalacticChimp commented on objects/obj_mass_equip/Step_0.gml:
```diff
@@ -176,31 +268,121 @@ if (refresh=true) and (obj_controller.settings>0){// Refreshing
-                    if (hue2=0){hue=0;repeat(30){if (hue2=0){hue+=1;if (we1[hue]=""){hue2=-5;we1[hue]=obj_ini.wep1[co,i];we1_n[hue]=1;}}}}
+                if (obj_ini.wep1[co,i]!=""){
+					hue=0;hue2=0;
+					for(hue=0;hue<60;hue++){
```
same, `hue=1; hue <= 60`

GalacticChimp reviewed (changes requested):
please check my comments + take a look at all modified loops, whether they have proper params

felipeStouvenel commented:
Fixed issue, was regarding starting at -1; updated to count at 1 instead of -1; the repeat automatically added 1 at the beggining of the loop, so thats why there was a mistake.
![image](https://github.com/ZAB909/ChapterMaster/assets/142210479/bd6766f4-b350-4aef-b336-5979906ad40c)

GalacticChimp commented:
shop still has arabic numerals for companies:

![image](https://github.com/ZAB909/ChapterMaster/assets/63882831/6054541d-8ade-4e39-a344-876e20e580fb)

GalacticChimp commented:
apart from the shop, there are 2 other instances to change (although I don't know how to get to these views in-game):

![image](https://github.com/ZAB909/ChapterMaster/assets/63882831/6f57f9f3-debd-4226-8b1c-e757c50fb987)

GalacticChimp reviewed (changes requested):
please modify remaining company numeral instances

felipeStouvenel commented:
Fixed all instances where company was being used, refactored a couple more scripts (mainly beautify) and corrected some erros from the GM console (red ones)
![image](https://github.com/ZAB909/ChapterMaster/assets/142210479/190ee614-3840-43fc-b57e-ab31e0becdea)

GalacticChimp commented on objects/obj_mass_equip/Step_0.gml:
```diff
@@ -72,81 +87,159 @@ if (engage=true){
-    repeat(11){co+=1;i=0;
-        repeat(300){i+=1;
+	
+	for(var co=1; co<11; co++){
```
`co == 0` (old code had `co=-1` at start)

GalacticChimp commented on objects/obj_mass_equip/Step_0.gml:
```diff
@@ -228,8 +413,8 @@ if (refresh=true) and (obj_controller.settings>0){// Refreshing
-    var i;i=-1;
-    repeat(11){i+=1;
+	
+	for(var i=1; i<=11; i++){
```
`i == 0; i < 11`, old code had `i = -1`

GalacticChimp commented on objects/obj_popup/Draw_0.gml:
```diff
@@ -138,8 +146,8 @@ if ((zm=0) and (type<=4)) or (type=98){
     // if (text2!="") then draw_text_ext(x1+126,y1+309,string(text2),-1,384);
-    
-    var img;img=-1;
+    // TODO change this into an array in a function (like romanNumerals does in here)
```
for a future PR?

felipeStouvenel replied:
yes, basically would add even more work. My idea is to basically have a simple structure:
something like 
{
    enemies: {
        tau:{ 
            image: 1;
            enemy: 1;
            other stuff 
        }
    }
}

basically have a json with all the enemy data, and move into json structures rather than simple arrays or just strings. But that  is another PR imo

GalacticChimp commented on objects/obj_popup/Draw_0.gml:
```diff
@@ -685,55 +741,98 @@ if (zm=0) and (type=6) and (instance_exists(obj_controller)){// Changing Equipme
+        if (min_exp<80) then draw_set_alpha(0.25);
+        // I Company
+        if (target_comp=1) then check="x";
+        draw_text(xx+1030,yy+230,string_hash_to_newline(romanNumerals[target_comp-1]+" ["+string(check)+"]"));
```
you could simplify all this code by extracting a method, something like this:

```
function draw_company_text(x_pos., min_exp_threshold){
    draw_text(x_pos, ,yy+230,string_hash_to_newline(romanNumerals[target_comp-1]+" ["+x+"]"));
    draw_set_alpha(1);
    if (min_exp<min_exp_threshold) then draw_set_alpha(0.25);
}
```

GalacticChimp replied:
as we discussed, can be done in separate PR

GalacticChimp commented on objects/obj_popup/Draw_0.gml:
```diff
@@ -834,55 +958,83 @@ if (zm=0) and (type=5) and (instance_exists(obj_controller)){// Promoting
-    if (target_comp=8) then check="x";draw_text(xx+1250,yy+250,string_hash_to_newline("8th ["+string(check)+"]"));check=" ";
-    if (target_comp=9) then check="x";draw_text(xx+1360,yy+250,string_hash_to_newline("9th ["+string(check)+"]"));check=" ";
-    if (target_comp=10) then check="x";draw_text(xx+1470,yy+250,string_hash_to_newline("10th ["+string(check)+"]"));check=" ";}
+        // I Company
```
can extract a method, similar as above

GalacticChimp replied:
like we discussed, this can be done in another PR

GalacticChimp commented on objects/obj_shop/Draw_0.gml:
```diff
@@ -10,8 +11,8 @@ draw_set_color(c_gray);
-var te;te="";if (target_comp=1) then te="1st";if (target_comp=2) then te="2nd";
-if (target_comp=3) then te="3rd";if (target_comp>3) then te=string(target_comp)+"th";
+var te="";
+if (target_comp<=10) then te=romanNumerals[target_comp-1];
```
could this be ever >10? and if so, what should be done here?

felipeStouvenel replied:
So far I dont think so, but i cant find where target_comp is defined. If its more than 10 probably its HQ or one of the others non company organizations in the panel. Need to see more of the logic. Will mark as a TODO for now and we could add a new PR for how the company is defined?

GalacticChimp replied:
ok, mark as TODO

GalacticChimp commented on scripts/scr_company_view/scr_company_view.gml:
```diff
@@ -46,30 +73,31 @@ function scr_company_view(argument0) {
-	                        if (sel_uni[i]="") and (op=0) then op=i;
-	                        if (sel_uni[i]=ma_role[v]) then go=1;
-	                    }
+	                    for (var j=0; j<=20;j++) {
```
`j < 20`

GalacticChimp commented on scripts/scr_company_view/scr_company_view.gml:
```diff
@@ -46,30 +73,31 @@ function scr_company_view(argument0) {
-	                    repeat(20){i+=1;
-	                        if (sel_veh[i]="") and (op=0) then op=i;
-	                        if (sel_veh[i]=ma_role[v]) then go=1;
+	                    for (var j=0; j<=20;j++) {
```
`j < 20`

GalacticChimp commented on scripts/scr_management/scr_management.gml:
```diff
@@ -3,185 +3,340 @@ function scr_management(argument0) {
+			obj_managment_panel.italic[q]=1;
+		}
+		
+		// TODO add specific Space Wolves and successor chapter logic for Master of Sanctity
```
for a later PR?

felipeStouvenel replied:
Yes, it would need a whole new refactor of a lot of code, since SW have a ton of special names and they handle chaplains and apothecaries differently from what i recal from lore. Same type of PR would be needed for say Iron Hands

GalacticChimp commented on scripts/scr_ui_advisors/scr_ui_advisors.gml:
```diff
@@ -1,19 +1,61 @@
+    	"moderate",
+    	"fast",
+    	"frenetic",
+    	"hereticly fast"
```
if these are not localisation strings but string ids, then I highly recommend using _ instead of spaces, will make our life easier once we switch to json

GalacticChimp commented on scripts/scr_ui_advisors/scr_ui_advisors.gml:
```diff
@@ -463,15 +547,22 @@ function scr_ui_advisors() {
             
-	            fest_type="Great Feast";fest_star="";
-	            fest_sid=0;fest_wid=0;fest_planet=0;
+	            fest_type="Great Feast";
```
fe**a**st_type

GalacticChimp replied:
ok, festivity is fine

GalacticChimp commented on scripts/scr_ui_advisors/scr_ui_advisors.gml:
```diff
@@ -513,110 +605,218 @@ function scr_ui_advisors() {
 	        }
         
-	        // Attendants
+	        // TODO Attendants 
```
future PR?

felipeStouvenel replied:
Another PR would ideally be the whole restructure of the celebrations/parades you do in the chapter. But would need to re read lore to see how this stuff was handled and probably varies with each chapter somewhat

GalacticChimp commented on scripts/scr_ui_advisors/scr_ui_advisors.gml:
```diff
@@ -908,33 +1162,36 @@ function scr_ui_advisors() {
+		blurp="";
 	    max_techs=round((disposition[3]/2))+5;
-    
-	    // show_message("Max Techmarines: "+string(max_techs)+"   Have: "+string(temp[37]));
```
👍

GalacticChimp reviewed (changes requested):
added some comments

GalacticChimp commented on objects/obj_mass_equip/Step_0.gml:
```diff
@@ -72,81 +87,159 @@ if (engage=true){
+    total_role_number=0;
+	total_roles="";
+	for(var i=1; i<=11; i++){role_number[i]=0;}
+	for(var i=1; i<=61; i++){
```
`i = 0`

GalacticChimp commented on scripts/scr_company_view/scr_company_view.gml:
```diff
@@ -239,45 +270,44 @@ function scr_company_view(argument0) {
+	                var go=0,op=0;
 	                if (man[v]="vehicle"){
-	                    repeat(20){p+=1;
+	                    for (var p=1;p<=20;p++){
```
`p = 0; p < 20`

GalacticChimp reviewed (changes requested):
added 3 new comments, but apart from them the PR looks fine. After you resolve the 3 comments, I'll pull locally, test, then approve and merge.

felipeStouvenel commented:
Fixed the error for transfer and promoting:
![image](https://github.com/ZAB909/ChapterMaster/assets/142210479/0e799b73-fd02-4fee-91fe-d94caa250433)
![image](https://github.com/ZAB909/ChapterMaster/assets/142210479/3fd280f0-a5d2-449f-9fc0-f86c506cc2fa)

felipeStouvenel commented:
Resolved merge conflicts

GalacticChimp commented:
> Resolved merge conflicts

looks like there are new conflicts

felipeStouvenel commented:
> > Resolved merge conflicts
> 
> looks like there are new conflicts

Re resolved conflicts

GalacticChimp commented:
weapons do not render:

![image](https://github.com/ZAB909/ChapterMaster/assets/63882831/1eb714e1-9fc6-4c12-9308-8c021b5d4e65)

also the game doesn't compile:

![image](https://github.com/ZAB909/ChapterMaster/assets/63882831/d108220f-b1bf-4616-812d-2abb85faa032)

GalacticChimp reviewed (changes requested):
please check my newest comment

felipeStouvenel commented:
Fixed:
![image](https://github.com/ZAB909/ChapterMaster/assets/142210479/31c59685-f71b-4664-9f16-54456a9891ed)
